neural_lam/weather_dataset.py

```diff
@@ -94,6 +126,66 @@ def __init__(
                     "the data in `BaseDatastore.get_dataarray`?"
                 )
 
+        # handling ensemble data
+        if self.datastore.is_ensemble:
+            # for the now the strategy is to only include the first ensemble
+            # member
+            # XXX: this could be changed to include all ensemble members by
+            # splitting `idx` into two parts, one for the analysis time and one
+            # for the ensemble member and then increasing self.__len__ to
+            # include all ensemble members
+            warnings.warn(
+                "only use of ensemble member 0 (the first member) is "
+                "implemented for ensemble data"
+            )
+            i_ensemble = 0
+            self.da_state = self.da_state.isel(ensemble_member=i_ensemble)
+        else:
+            self.da_state = self.da_state
+
+        # Check time step consistency in state data and determine time steps
+        # for state, forcing and boundary data
+        if self.datastore.is_forecast:
+            state_times = self.da_state.analysis_time
+            self.forecast_step_state = get_time_step(
+                self.da_state.elapsed_forecast_duration
+            )
+        else:
+            state_times = self.da_state.time
+        self.time_step_state = get_time_step(state_times)
+        if self.da_forcing is not None:
+            if self.datastore.is_forecast:
+                forcing_times = self.da_forcing.analysis_time
+                self.forecast_step_forcing = self._get_time_step(
+                    self.da_forcing.elapsed_forecast_duration
+                )
+            else:
+                forcing_times = self.da_forcing.time
+            self.time_step_forcing = self._get_time_step(forcing_times.values)
+            if self.datastore_boundary.is_forecast:
+                boundary_times = self.da_boundary_forcing.analysis_time
+                self.forecast_step_boundary = self._get_time_step(
+                    self.da_boundary_forcing.elapsed_forecast_duration
+                )
+            else:
+                boundary_times = self.da_boundary_forcing.time
+            self.time_step_boundary = self._get_time_step(boundary_times.values)
+
+        # Forcing data is part of the same datastore as state data
+        # During creation the time dimension of the forcing data
+        # is matched to the state data
+        # Boundary data is part of a separate datastore
+        # The boundary data is allowed to have a different time_step
+        # Check that the boundary data covers the required time range
+        check_time_overlap(
+            self.da_state,
+            self.da_boundary_forcing,
+            da1_is_forecast=self.datastore.is_forecast,
+            da2_is_forecast=self.datastore_boundary.is_forecast,
+            num_past_steps=self.num_past_boundary_steps,
+            num_future_steps=self.num_future_boundary_steps,
+        )
```

Comment:
Joel Oskarsson
  Tuesday at 2:16 PM
I've been trying out loading data in neural-lam that I have created using the DANRA and ERA5 configs from https://github.com/leifdenby/mllam-data-prep/tree/feat/crop-with-other-dataset. In those configs both datasets cover exactly the same time period. Running neural-lam with num_past_boundary_steps > 0 raises this error https://github.com/sadamov/neural-lam/blob/f0a7046b7ce0c5c955762b67ebe017fdef4eae4e/neural_lam/weather_dataset.py#L207 (even with the fix https://github.com/mllam/neural-lam/pull/90#discussion_r1888481106). This makes sense, as there is no boundary data before the first time point for the interior. However, an alternative behavior would be to adjust the first time point for the interior (e.g. ignoring the first time steps of the interior data, until data loading is possible for the set num_past_boundary_steps ), to still allow for training in this case. What's your thoughts on this? Now one has to always leave in some extra time steps in the boundary forcing, depending on num_past_boundary_steps , which has to be decided beforehand when the dataset is created. But maybe that makes more sense than allowing for having a different number of samples depending on num_past_boundary_steps ?
------------------------------------------------------
Simon Adamov:
true this is indeed by design.
tell the user to include sufficient time-steps and have a very clear requirement and warning
OR: expect most users to chose the same datetimes for boundary and forcing and give a warning how many time steps were removed from interior
As long as only a few timesteps are removed from interior, I think option 2 is nicer to handle. But no user would want half their training data to be ignored "quietly".
So, I think both options are fine, as long as we have clear warnings and error messages. Is 2 your preference?
----------------------------------------------------------------
Joel Oskarsson
  Wednesday at 5:30 PM
That's good ideas. I agree that we should really avoid cutting away much of the training data without any proper warning. But cutting away a few timesteps is not really a problem (but could still show a warning explaining how much is being removed). Yea, option 2 is quite natural I think, since it would be nice to prepare the data by just specifying the exact same time period for the interior and boundary when building corresponding zarrs :grin: I'm thinking that this might be easy to achieve by "just" slicing away a few time steps in the interior xr.Dataarray used in WeatherDataset, but not I have not looked at any details of what such an implementation would look like.

Reply:
I suggest to implement an additional stateless function called `crop_time_if_needed` in 7e5797e18f5a251441e98ef9ad63f412c06ef409. This function gets called automatically before any time-slicing operations in WeatherDataset. I left all the existing checks and warning in the code, can't hurt. Added additional print statement about how many time steps were removed from the state data. Let me know what you think!